File: ios/VoiceCode/Managers/LogManager.swift
// LogManager.swift
// Captures and manages app logs for debugging

import Foundation

class LogManager {
    static let shared = LogManager()

    private let maxLogLines = 5000 // Keep last 5000 lines in memory
    private var logLines: [String] = []
    private let queue = DispatchQueue(label: "com.travisbrown.VoiceCode.LogManager")

    private init() {
        // Start capturing logs
        startCapturing()
    }

    private func startCapturing() {
        // Logs are captured via explicit LogManager.shared.log() calls
        // throughout the app; there is no automatic redirection here.
    }

    /// Append a log message (call this from your logging points)
    func log(_ message: String, category: String = "General") {
        queue.async {
            let timestamp = DateFormatter.logTimestamp.string(from: Date())
            let logLine = "[\(timestamp)] [\(category)] \(message)"
            self.logLines.append(logLine)

            // Keep only last N lines
            if self.logLines.count > self.maxLogLines {
                self.logLines.removeFirst(self.logLines.count - self.maxLogLines)
            }
        }
    }

    /// Get the last N lines of logs (default 100KB worth, complete lines only)
    func getRecentLogs(maxBytes: Int = 100_000) -> String {
        return queue.sync {
            // Join all lines
            let allLogs = logLines.joined(separator: "\n")

            // If total size is under limit, return all
            guard let data = allLogs.data(using: .utf8), data.count > maxBytes else {
                return allLogs
            }

            // Find the cutoff point to stay under maxBytes
            // Start from the end and work backwards
            var totalBytes = 0
            var includedLines: [String] = []

            for line in logLines.reversed() {
                let lineData = (line + "\n").data(using: .utf8) ?? Data()
                if totalBytes + lineData.count > maxBytes {
                    break
                }
                totalBytes += lineData.count
                includedLines.insert(line, at: 0)
            }

            return includedLines.joined(separator: "\n")
        }
    }

    /// Get all captured logs
    func getAllLogs() -> String {
        return queue.sync {
            logLines.joined(separator: "\n")
        }
    }

    /// Clear all logs
    func clearLogs() {
        queue.async {
            self.logLines.removeAll()
        }
    }
}

// MARK: - Date Formatter Extension

extension DateFormatter {
    static let logTimestamp: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss.SSS"
        return formatter
    }()
}

File: android/app/src/main/kotlin/dev/labs910/voicecode/util/LogManager.kt
package dev.labs910.voicecode.util

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Captures and manages app logs for debugging.
 * Corresponds to iOS LogManager.swift.
 *
 * Key features:
 * - Thread-safe log capture with coroutine mutex
 * - Configurable max lines (default 1000)
 * - Size-limited retrieval for efficient transmission
 * - Timestamp formatting consistent with iOS
 */
object LogManager {

    private const val MAX_LOG_LINES = 1000
    private const val DEFAULT_MAX_BYTES = 15_000

    private val logLines = CopyOnWriteArrayList<String>()
    private val mutex = Mutex()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val timestampFormat = SimpleDateFormat("HH:mm:ss.SSS", Locale.US).apply {
        timeZone = TimeZone.getDefault()
    }

    /**
     * Append a log message.
     *
     * @param message The log message
     * @param category Log category (default "General")
     */
    fun log(message: String, category: String = "General") {
        scope.launch {
            mutex.withLock {
                val timestamp = timestampFormat.format(Date())
                val logLine = "[$timestamp] [$category] $message"
                logLines.add(logLine)

                // Keep only last N lines
                while (logLines.size > MAX_LOG_LINES) {
                    logLines.removeAt(0)
                }
            }
        }
    }

    /**
     * Convenience method for logging with different levels.
     */
    fun debug(message: String, category: String = "Debug") = log(message, category)
    fun info(message: String, category: String = "Info") = log(message, category)
    fun warn(message: String, category: String = "Warning") = log(message, category)
    fun error(message: String, category: String = "Error") = log(message, category)

    /**
     * Get the last N bytes worth of logs (complete lines only).
     *
     * @param maxBytes Maximum bytes to return (default 15KB)
     * @return Recent log lines joined by newlines
     */
    suspend fun getRecentLogs(maxBytes: Int = DEFAULT_MAX_BYTES): String {
        return withContext(Dispatchers.IO) {
            mutex.withLock {
                val allLogs = logLines.joinToString("\n")

                // If total size is under limit, return all
                val totalBytes = allLogs.toByteArray(Charsets.UTF_8).size
                if (totalBytes <= maxBytes) {
                    return@withLock allLogs
                }

                // Find the cutoff point to stay under maxBytes
                // Start from the end and work backwards
                var currentBytes = 0
                val includedLines = mutableListOf<String>()

                for (line in logLines.reversed()) {
                    val lineBytes = (line + "\n").toByteArray(Charsets.UTF_8).size
                    if (currentBytes + lineBytes > maxBytes) {
                        break
                    }
                    currentBytes += lineBytes
                    includedLines.add(0, line)
                }

                includedLines.joinToString("\n")
            }
        }
    }

    /**
     * Get the last N bytes worth of logs synchronously (for non-coroutine contexts).
     * Less efficient than the suspend version - use suspend version when possible.
     *
     * @param maxBytes Maximum bytes to return (default 15KB)
     * @return Recent log lines joined by newlines
     */
    fun getRecentLogsSync(maxBytes: Int = DEFAULT_MAX_BYTES): String {
        val allLogs = logLines.joinToString("\n")

        // If total size is under limit, return all
        val totalBytes = allLogs.toByteArray(Charsets.UTF_8).size
        if (totalBytes <= maxBytes) {
            return allLogs
        }

        // Find the cutoff point to stay under maxBytes
        var currentBytes = 0
        val includedLines = mutableListOf<String>()

        for (line in logLines.reversed()) {
            val lineBytes = (line + "\n").toByteArray(Charsets.UTF_8).size
            if (currentBytes + lineBytes > maxBytes) {
                break
            }
            currentBytes += lineBytes
            includedLines.add(0, line)
        }

        return includedLines.joinToString("\n")
    }

    /**
     * Get all captured logs.
     *
     * @return All log lines joined by newlines
     */
    suspend fun getAllLogs(): String {
        return withContext(Dispatchers.IO) {
            mutex.withLock {
                logLines.joinToString("\n")
            }
        }
    }

    /**
     * Get all captured logs synchronously.
     *
     * @return All log lines joined by newlines
     */
    fun getAllLogsSync(): String {
        return logLines.joinToString("\n")
    }

    /**
     * Clear all logs.
     */
    fun clearLogs() {
        scope.launch {
            mutex.withLock {
                logLines.clear()
            }
        }
    }

    /**
     * Clear all logs synchronously.
     */
    fun clearLogsSync() {
        logLines.clear()
    }

    /**
     * Get the current number of log lines.
     *
     * @return Number of stored log lines
     */
    fun getLogCount(): Int = logLines.size
}
